staging/src/k8s.io/component-base/zpages/statusz/statusz.go

```diff
@@ -0,0 +1,163 @@
+
+	"k8s.io/klog/v2"
+)
+
+var (
+	delimiters              = []string{":", ": ", "=", " "}
+	errUnsupportedMediaType = fmt.Errorf("media type not acceptable, must be: text/plain")
+)
+
+const (
+	headerFmt = `
+%s statusz
+Warning: This endpoint is not meant to be machine parseable, has no formatting compatibility guarantees and is for debugging purposes only.
+`
+
+	dataTemplate = `
+Started{{.Delim}} {{.StartTime}}
+Up{{.Delim}} {{.Uptime}}
+Go version{{.Delim}} {{.GoVersion}}
+Binary version{{.Delim}} {{.BinaryVersion}}
+Emulation version{{.Delim}} {{.EmulationVersion}}
+Minimum Compatibility version{{.Delim}} {{.MinCompatibilityVersion}}
+`
+)
+
+type contentFields struct {
+	Delim                   string
+	StartTime               string
+	Uptime                  string
+	GoVersion               string
+	BinaryVersion           string
+	EmulationVersion        string
+	MinCompatibilityVersion string
+}
+
+type mux interface {
+	Handle(path string, handler http.Handler)
+}
+
+func NewRegistry() statuszRegistry {
+	return registry{}
+}
+
+func Install(m mux, componentName string, reg statuszRegistry) {
+	dataTmpl, err := initializeTemplates()
+	if err != nil {
+		klog.Errorf("error while parsing gotemplates: %v", err)
+		return
+	}
+	m.Handle("/statusz", handleStatusz(componentName, dataTmpl, reg))
+}
+
+func initializeTemplates() (*template.Template, error) {
+	d := template.New("data")
+	dataTmpl, err := d.Parse(dataTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	return dataTmpl, nil
+}
+
+func handleStatusz(componentName string, dataTmpl *template.Template, reg statuszRegistry) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if !acceptableMediaType(r) {
+			http.Error(w, errUnsupportedMediaType.Error(), http.StatusNotAcceptable)
+			return
+		}
+
+		fmt.Fprintf(w, headerFmt, componentName)
+		data, err := populateStatuszData(dataTmpl, reg)
+		if err != nil {
+			klog.Errorf("error while populating statusz data: %v", err)
+			http.Error(w, "error while populating statusz data", http.StatusInternalServerError)
+			return
+		}
+
+		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+		fmt.Fprint(w, data)
+	}
+}
+
+func acceptableMediaType(r *http.Request) bool {
+	accepts := goautoneg.ParseAccept(r.Header.Get("Accept"))
+	for _, accept := range accepts {
+		if !mediaTypeMatches(accept) {
+			continue
+		}
+		if len(accept.Params) == 0 {
+			return true
+		}
+		if len(accept.Params) == 1 {
+			if charset, ok := accept.Params["charset"]; ok && strings.EqualFold(charset, "utf-8") {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func mediaTypeMatches(a goautoneg.Accept) bool {
+	return (a.Type == "text" || a.Type == "*") &&
+		(a.SubType == "plain" || a.SubType == "*")
+}
+
+func populateStatuszData(tmpl *template.Template, reg statuszRegistry) (string, error) {
+	if tmpl == nil {
+		return "", fmt.Errorf("received nil template")
+	}
+
+	randomIndex := rand.Intn(len(delimiters))
```

Comment:
What's the motivation to prevent the page from being machine parsed? Is it because the page could be changed in the future or that automation shouldn't access this page at all?

Reply:
> Is it because the page could be changed in the future and has no formatting guarantees or that automation shouldn't access this page at all?

yeah this was @liggitt's suggestion to ensure that users understand that we are deliberately making this endpoint un-stable for alpha (till we have a proper path forward toward having a stable API for z-endpoints)

Reply:
When someone does parse this and we then break them. We'll point them to the warning?  Is there any other documentation about the non-backward compatibility here or just the warning?

Reply:
Just the warning returned in the payload and in the KEP